File: number-of-connected-components-in-an-undirected-graph.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int countComponents(int n, vector<vector<int, int>> edges) {
        unordered_map<int, vector<int>> graph;
        
        for (int i=0; i<n; ++i) {
            graph[i] = vector<int>();
        }

        for (vector<int> v: edges) {
            graph[v[0]].push_back(v[1]);
            graph[v[1]].push_back(v[0]);
        }

        unordered_set<int> visited;
        int count = 0;
        for (auto p: graph) {
            if (visited[p.first]) {
                continue;
            }
            count += 1;
            visit(p.first, -1, graph, visited);
        }
        return count;
    }

    void visit(int node, int prev, unordered_map<int, vector<int>>& graph, unordered_set<int>& visited) {
        if (visited.find(node) != visited.end()) {
            return;
        }

        visited.insert(node);
        for (int adj: graph[node]) {
            visit(adj, node, graph, visited)
        }

        return true;
    }
};

File: number-of-connected-components-in-an-undirected-graph.py
from typing import List

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = {i: [] for i in range(n)}

        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)

        visited = [False for _ in range(n)]

        def visit(node: int, prev: int):
            if visited[node]:
                return

            visited[node] = True
            for adj in graph[node]:
                visit(adj, node)

        count = 0
        for i, adj_list in graph.items():
            if visited[i]:
                continue

            count += 1
            visit(i, -1)

        return count
